src/transformers/trainer.py

```diff
@@ -595,27 +595,29 @@ def __init__(
             if args.half_precision_backend == "cuda_amp":
                 self.use_cuda_amp = True
                 self.amp_dtype = torch.float16 if args.fp16 else torch.bfloat16
-                self.do_grad_scaling = True
-                if self.sharded_ddp is not None:
-                    self.scaler = ShardedGradScaler()
-                elif self.fsdp is not None:
-                    if self.amp_dtype == torch.float16:
-                        from torch.distributed.fsdp.sharded_grad_scaler import (
-                            ShardedGradScaler as FSDPShardedGradScaler,
-                        )
+                #  bf16 does not need grad scaling
+                self.do_grad_scaling = self.amp_dtype == torch.float16
+                if self.do_grad_scaling:
+                    if self.sharded_ddp is not None:
+                        self.scaler = ShardedGradScaler()
+                    elif self.fsdp is not None:
+                        if self.amp_dtype == torch.float16:
+                            from torch.distributed.fsdp.sharded_grad_scaler import (
+                                ShardedGradScaler as FSDPShardedGradScaler,
+                            )
 
-                        self.scaler = FSDPShardedGradScaler()
-                    else:
-                        self.do_grad_scaling = False
-                        self.use_cuda_amp = False
-                        self.amp_dtype = None
+                            self.scaler = FSDPShardedGradScaler()
+                        else:
+                            self.do_grad_scaling = False
+                            self.use_cuda_amp = False
+                            self.amp_dtype = None
```

Comment:
Just realized there is this else block here. Clearly `self.do_grad_scaling = False` is not necessary, but you might need to have the two other lines somewhere else.

@pacman100 FSDP doesn't handle bfloat16 at all?

Reply:
Hello @sgugger, similar to DeepSpeed, FSDP also manages their own half-precision, however, for FP16 it needs `ShardedGradScaler`.  Here's an example notebook from PyTorch team wrt FSDP MixedPrecision: https://github.com/lessw2020/transformer_central/blob/main/mixed_precision/mixed_precision_fsdp.ipynb